### PR TITLE
[WIP] Native notifications

### PR DESCRIPTION
**Pull request by vivangkumar**
Adds native notification support to the library.

*DO NOT MERGE THIS YET*

TODO:
- ~~README~~
- ~~Comments/ Docs~~
- ~~Tests~~
- ~~Update endpoint~~

@zimbatm

### CODE REVIEW

File: lib/Pusher.php [hunk 1]
```diff
@@ -268,9 +281,13 @@ private function create_curl($s_url, $request_method = 'GET', $query_params = ar
             $s_url,
             $query_params);
 
-        $full_url = $this->settings['scheme'].'://'.
+        if ($notification_host) {
+            $full_url = 'https://'.$this->settings['notification_host'].$s_url.'?'.$signed_query;
```
Comment (zimbatm) [inline lib/Pusher.php]: what if the user wants to use plain http? (like for testing on a mock server)

File: lib/Pusher.php [hunk 2]
```diff
@@ -258,8 +269,10 @@ private function validate_socket_id($socket_id)
     /**
      * Utility function used to create the curl object with common settings.
      */
-    private function create_curl($s_url, $request_method = 'GET', $query_params = array())
+    private function create_curl($s_url, $request_method = 'GET', $query_params = array(), $notification_host = false)
```
Comment (zimbatm) [inline lib/Pusher.php]: I think we need a function that takes a base_url such as `$base_url . $s_url .'?'.$signed_query` gives the full url. Then introduce a `create_notification_curl` and `create_ddn_curl` wrapper functions that pass the right base_url.

Comment (vivangkumar) [inline lib/Pusher.php]: Fair. I didn't want to change things much, but that sounds like a much much better way.

File: lib/Pusher.php (hunk 1 above)
Comment (vivangkumar) [inline lib/Pusher.php]: Perhaps I can introduce another option to specify the notification scheme.

File: lib/Pusher.php [hunk 3]
```diff
@@ -147,6 +153,20 @@ public function __construct($auth_key, $secret, $app_id, $options = array(), $ho
             }
         }
 
+        // Set the native notification host
+        if (isset($options['notification_host'])) {
```
Comment (zimbatm) [inline lib/Pusher.php]: you can use `if($options['notification_host'])`. In that case the empty string will be False and use the default.

File: lib/Pusher.php [hunk 4]
```diff
@@ -147,6 +153,20 @@ public function __construct($auth_key, $secret, $app_id, $options = array(), $ho
...
+            $this->settings['notification_host'] = $options['notification_host'];
+        } else {
+            $this->settings['notification_host'] = 'hedwig-staging.herokuapp.com';
+        }
+
+        // Set scheme for native notifications
+        if (isset($options['notification_scheme'])) {
```
Comment (zimbatm) [inline lib/Pusher.php]: Same

File: lib/Pusher.php [hunk 5]
```diff
@@ -331,6 +351,88 @@ private function exec_curl($ch)
...
+     *
+     * @throws PusherException if validation fails
+     */
+    private function validate_notification_payload($payload)
+    {
+        if (!(isset($payload['gcm']) || isset($payload['apns']))) {
+            throw new PusherException('GCM or APNS data must be provided');
+        }
+
+        if (isset($payload['gcm'])) {
```
Comment (zimbatm) [inline lib/Pusher.php]: Same

File: lib/Pusher.php (hunk 3 above)
Comment (vivangkumar) [inline lib/Pusher.php]: Doesn't isset do that same? it also checks for null values, right?

**Pull request merged**